File: timer_manager.cpp
// timer_manager.cpp
#include <ncurses.h>
#include <json.hpp>
#include <fstream>
#include <vector>
#include <string>
#include <unistd.h>
#include <chrono>
#include <thread>

using json = nlohmann::json;

struct Timer {
    std::string name;
    double duration; // seconds
    double elapsed;
    bool running;
    bool finished;

    double remaining() const {
        if (finished) return 0;
        double rem = duration - elapsed;
        return rem > 0 ? rem : 0;
    }

    bool update(double dt) {
        if (running && !finished) {
            elapsed += dt;
            if (elapsed >= duration) {
                elapsed = duration;
                running = false;
                finished = true;
                return true;
            }
        }
        return false;
    }

    json to_json() const {
        return {{"name", name}, {"duration", duration}, {"elapsed", elapsed}, {"running", running}, {"finished", finished}};
    }

    static Timer from_json(const json& j) {
        return {j["name"], j["duration"], j["elapsed"], j["running"], j["finished"]};
    }
};

class TimerManager {
public:
    std::vector<Timer> timers;
    int selected;
    bool running;
    std::string saveFile;

    TimerManager(const std::string& save) : selected(0), running(true), saveFile(save) {
        load();
        if (timers.empty()) {
            timers.push_back({"Pomodoro", 25*60, 0, false, false});
            timers.push_back({"Break", 5*60, 0, false, false});
        }
    }

    void addTimer(const std::string& name, double duration) {
        timers.push_back({name, duration, 0, false, false});
        save();
    }

    void removeTimer(int idx) {
        if (idx < 0 || idx >= (int)timers.size()) return;
        timers.erase(timers.begin() + idx);
        if (selected >= (int)timers.size()) selected = timers.size()-1;
        save();
    }

    void toggle(int idx) {
        if (idx < 0 || idx >= (int)timers.size()) return;
        Timer& t = timers[idx];
        if (t.finished) {
            t.elapsed = 0;
            t.finished = false;
            t.running = true;
        } else {
            t.running = !t.running;
        }
        save();
    }

    void reset(int idx) {
        if (idx < 0 || idx >= (int)timers.size()) return;
        Timer& t = timers[idx];
        t.elapsed = 0;
        t.running = false;
        t.finished = false;
        save();
    }

    void update(double dt) {
        for (auto& t : timers) {
            if (t.update(dt)) {
                mvprintw(0, 0, "!!! Timer '%s' finished !!!", t.name.c_str());
            }
        }
    }

    void save() {
        json j = json::array();
        for (const auto& t : timers) j.push_back(t.to_json());
        std::ofstream file(saveFile);
        file << j.dump(4);
    }

    void load() {
        std::ifstream file(saveFile);
        if (file.is_open()) {
            json j;
            file >> j;
            for (const auto& item : j) {
                timers.push_back(Timer::from_json(item));
            }
        }
    }

    void run() {
        initscr();
        cbreak();
        noecho();
        keypad(stdscr, TRUE);
        curs_set(0);
        nodelay(stdscr, TRUE);

        auto last = std::chrono::steady_clock::now();
        int ch;
        while (running) {
            clear();
            mvprintw(0, 0, "=== Timer Manager ===");
            mvprintw(1, 0, "Timers: %zu   [a]dd  [d]elete  [Space]toggle  [r]eset  [q]uit", timers.size());

            int line = 3;
            for (int i=0; i<(int)timers.size(); i++) {
                Timer& t = timers[i];
                double rem = t.remaining();
                const char* status = t.finished ? " ✓" : (t.running ? " ▶" : " ⏸");
                char buf[100];
                sprintf(buf, "%d. %s %s  %02.0f:%02.0f / %02.0f:%02.0f",
                        i+1, t.name.c_str(), status,
                        rem/60, fmod(rem,60),
                        t.duration/60, fmod(t.duration,60));
                if (i == selected) attron(A_REVERSE);
                else {
                    int color = t.running ? COLOR_GREEN : (t.finished ? COLOR_RED : COLOR_YELLOW);
                    attron(COLOR_PAIR(color));
                }
                mvprintw(line, 0, "%s", buf);
                if (i == selected) attroff(A_REVERSE);
                else attroff(COLOR_PAIR(t.running ? COLOR_GREEN : (t.finished ? COLOR_RED : COLOR_YELLOW)));
                line++;
            }
            mvprintw(line+1, 0, "Commands: a=add, d=delete, Space=toggle, r=reset, q=quit");
            refresh();

            ch = getch();
            if (ch == 'q' || ch == 27) running = false;
            else if (ch == 'a') {
                echo();
                char name[100]; int dur;
                mvprintw(20, 0, "Enter name: ");
                getstr(name);
                mvprintw(21, 0, "Enter duration (seconds): ");
                scanw("%d", &dur);
                noecho();
                addTimer(name, dur);
            } else if (ch == 'd') {
                if (!timers.empty()) removeTimer(selected);
            } else if (ch == ' ') {
                if (!timers.empty()) toggle(selected);
            } else if (ch == 'r') {
                if (!timers.empty()) reset(selected);
            } else if (ch == KEY_UP) {
                if (selected > 0) selected--;
            } else if (ch == KEY_DOWN) {
                if (selected < (int)timers.size()-1) selected++;
            }

            auto now = std::chrono::steady_clock::now();
            double dt = std::chrono::duration<double>(now - last).count();
            last = now;
            update(dt);
            usleep(50000);
        }
        endwin();
        save();
    }
};

int main() {
    TimerManager tm("timers.json");
    tm.run();
    return 0;
}

File: timer_manager.py
# timer_manager.py
import curses
import json
import os
import time
import threading
import queue
from datetime import datetime

class Timer:
    def __init__(self, name, duration, elapsed=0, running=False):
        self.name = name
        self.duration = duration  # секунды
        self.elapsed = elapsed    # сколько прошло
        self.running = running
        self.finished = False

    def remaining(self):
        return max(0, self.duration - self.elapsed)

    def update(self, dt):
        if self.running and not self.finished:
            self.elapsed += dt
            if self.elapsed >= self.duration:
                self.elapsed = self.duration
                self.running = False
                self.finished = True
                return True  # сигнал о завершении
        return False

    def to_dict(self):
        return {
            'name': self.name,
            'duration': self.duration,
            'elapsed': self.elapsed,
            'running': self.running,
            'finished': self.finished
        }

    @classmethod
    def from_dict(cls, data):
        t = cls(data['name'], data['duration'], data['elapsed'], data['running'])
        t.finished = data.get('finished', False)
        return t

class TimerManager:
    def __init__(self, save_file='timers.json'):
        self.save_file = save_file
        self.timers = []
        self.selected = 0
        self.running = True
        self.last_update = time.time()
        self.notify_queue = queue.Queue()
        self.load()

    def add_timer(self, name, duration):
        self.timers.append(Timer(name, duration))
        self.save()

    def remove_timer(self, idx):
        if 0 <= idx < len(self.timers):
            del self.timers[idx]
            if self.selected >= len(self.timers):
                self.selected = len(self.timers) - 1
            self.save()

    def toggle(self, idx):
        if 0 <= idx < len(self.timers):
            t = self.timers[idx]
            if t.finished:
                # перезапуск
                t.elapsed = 0
                t.finished = False
                t.running = True
            else:
                t.running = not t.running
            self.save()

    def reset(self, idx):
        if 0 <= idx < len(self.timers):
            t = self.timers[idx]
            t.elapsed = 0
            t.running = False
            t.finished = False
            self.save()

    def update(self):
        now = time.time()
        dt = now - self.last_update
        self.last_update = now
        for i, timer in enumerate(self.timers):
            if timer.update(dt):
                self.notify_queue.put(timer.name)

    def save(self):
        data = [t.to_dict() for t in self.timers]
        with open(self.save_file, 'w') as f:
            json.dump(data, f, indent=2)

    def load(self):
        if os.path.exists(self.save_file):
            try:
                with open(self.save_file, 'r') as f:
                    data = json.load(f)
                    self.timers = [Timer.from_dict(d) for d in data]
            except:
                self.timers = []

def main(stdscr):
    curses.curs_set(0)
    stdscr.nodelay(1)
    curses.start_color()
    curses.init_pair(1, curses.COLOR_CYAN, curses.COLOR_BLACK)
    curses.init_pair(2, curses.COLOR_GREEN, curses.COLOR_BLACK)
    curses.init_pair(3, curses.COLOR_RED, curses.COLOR_BLACK)
    curses.init_pair(4, curses.COLOR_YELLOW, curses.COLOR_BLACK)

    manager = TimerManager()
    if not manager.timers:
        # Добавим пример
        manager.add_timer("Pomodoro", 25*60)
        manager.add_timer("Break", 5*60)

    input_buffer = ""
    input_mode = False
    new_name = ""
    new_duration = ""

    while manager.running:
        stdscr.clear()
        height, width = stdscr.getmaxyx()
        # Заголовок
        stdscr.addstr(0, 0, "=== Timer Manager ===", curses.A_BOLD)
        stdscr.addstr(1, 0, f"Timers: {len(manager.timers)}   [a]dd  [d]elete  [Space]toggle  [r]eset  [q]uit", curses.color_pair(1))

        # Список таймеров
        max_display = height - 6
        start = max(0, manager.selected - max_display//2)
        end = min(len(manager.timers), start + max_display)
        for i in range(start, end):
            t = manager.timers[i]
            remaining = t.remaining()
            status = " ▶" if t.running else " ⏸" if not t.finished else " ✓"
            if t.finished:
                status = " ✓"
            line = f"{i+1}. {t.name} {status}  {int(remaining//60):02d}:{int(remaining%60):02d} / {t.duration//60:02d}:{t.duration%60:02d}"
            if i == manager.selected:
                stdscr.addstr(i-start+2, 0, line, curses.A_REVERSE)
            else:
                color = curses.color_pair(2) if t.running else curses.color_pair(3) if t.finished else curses.color_pair(4)
                stdscr.addstr(i-start+2, 0, line, color)

        # Уведомления
        try:
            while True:
                name = manager.notify_queue.get_nowait()
                stdscr.addstr(height-3, 0, f"!!! Timer '{name}' finished !!!", curses.A_BOLD | curses.color_pair(3))
        except queue.Empty:
            pass

        # Подсказка внизу
        stdscr.addstr(height-2, 0, "Commands: a=add, d=delete, Space=toggle, r=reset, q=quit")
        if input_mode:
            stdscr.addstr(height-1, 0, f"Enter name: {new_name}  (press Enter to confirm)")
        stdscr.refresh()

        # Обработка ввода
        key = stdscr.getch()
        if key == ord('q') or key == 27:
            manager.save()
            break
        elif input_mode:
            if key == 10:  # Enter
                if new_name and new_duration:
                    try:
                        dur = int(new_duration)
                        manager.add_timer(new_name, dur)
                    except:
                        pass
                input_mode = False
                new_name = ""
                new_duration = ""
            elif key == 27:
                input_mode = False
                new_name = ""
                new_duration = ""
            elif key == ord('\t'):
                # переключение между name и duration
                pass
            elif key == curses.KEY_BACKSPACE or key == 127:
                if len(new_name) > 0:
                    new_name = new_name[:-1]
            else:
                if 32 <= key <= 126:
                    new_name += chr(key)
        else:
            if key == ord('a'):
                input_mode = True
                new_name = ""
                new_duration = ""
                stdscr.addstr(height-1, 0, "Enter timer name: ")
                curses.echo()
                name = stdscr.getstr().decode()
                curses.noecho()
                stdscr.addstr(height-1, 0, "Enter duration (seconds): ")
                curses.echo()
                dur_str = stdscr.getstr().decode()
                curses.noecho()
                try:
                    dur = int(dur_str)
                    manager.add_timer(name, dur)
                except:
                    pass
                # Обновим сразу
            elif key == ord('d'):
                if manager.timers:
                    manager.remove_timer(manager.selected)
            elif key == ord(' '):
                if manager.timers:
                    manager.toggle(manager.selected)
            elif key == ord('r'):
                if manager.timers:
                    manager.reset(manager.selected)
            elif key == curses.KEY_UP:
                manager.selected = max(0, manager.selected-1)
            elif key == curses.KEY_DOWN:
                manager.selected = min(len(manager.timers)-1, manager.selected+1)

        # Обновление таймеров
        manager.update()
        time.sleep(0.05)

if __name__ == "__main__":
    curses.wrapper(main)
